### PR TITLE
Fix gradient accumulation for Z2+offload

### PR DESCRIPTION
**Pull request by tohtana**
The ZeRO 1/2 optimizer performs incorrect gradient accumulation in the path for ZeRO2 + Offloading. This issue is caused by two main reasons:

1) The micro_step_id in the ZeRO 1/2 optimizer is:

- Initialized to 0 in the constructor.
- Reset to -1 during the backward pass.

For example, given a gradient accumulation step of 4, the micro_step_id changes as follows:

- For the first global step: 1, 2, 3, 4.
- Subsequently: 0, 1, 2, 3.

2) Gradients are copied to the buffer on the first micro step and accumulated in the buffer during the following micro steps. However, the current code incorrectly copies gradients at steps that are not at the accumulation boundary.

This PR aligns the micro_step_id initialization in both the constructor and the backward pass, and corrects the condition for copying and accumulating gradients.

### CODE REVIEW

**Review by tjruwase — APPROVED** [review]

**Comment by zwhe99** [comment]
Hi team,

I found that since this PR, the grad_norms of zero3 and zero2 are not in the same scale.

Here are some results of my testing:
```shell
# Before this PR

# zero2
{'loss': 1.0965, 'grad_norm': 28.10633659362793, 'learning_rate': 5e-06, 'epoch': 0.18}                                                                                                                                                             
{'loss': 1.1559, 'grad_norm': 29.9056339263916, 'learning_rate': 1e-05, 'epoch': 0.36} 

# zero3
{'loss': 1.0965, 'grad_norm': 28.104965209960938, 'learning_rate': 5e-06, 'epoch': 0.18}                                                                                                                                                            
{'loss': 1.1559, 'grad_norm': 29.90896987915039, 'learning_rate': 1e-05, 'epoch': 0.36}     
``` 

```shell
# After this PR

# zero2
{'loss': 1.0965, 'grad_norm': 3.760517120361328, 'learning_rate': 5e-06, 'epoch': 0.18}                                                                                                                                                             
{'loss': 1.1559, 'grad_norm': 4.289908409118652, 'learning_rate': 1e-05, 'epoch': 0.36}          

# zero3
{'loss': 1.0965, 'grad_norm': 28.104095458984375, 'learning_rate': 5e-06, 'epoch': 0.18}                                                                                                                                                            
{'loss': 1.1559, 'grad_norm': 29.90198516845703, 'learning_rate': 1e-05, 'epoch': 0.36}      
``` 

I'm wondering if this is as expected or if this is a bug.
